From 2a92e74e40dc0b93de70058d3fef421da9ebacee Mon Sep 17 00:00:00 2001
From: manukh <manukh@chromium.org>
Date: Mon, 16 Jun 2025 08:16:25 -0700
Subject: [PATCH] [omnibox] Fix crashes when classifying international text

Some international characters become multiple characters when converting
case. E.g. Armenian և is 1 character lowercase, but 2 characters Եվ
uppercased. Turkish has examples of the opposite, where the lowercasing
can increase length.

This was problematic when classifying match contents & description.
Classification uses the lowercased version of the user input and match
texts in order to match the input 'yOutuBE' with the match text
'YouTube'.

But if the input or match text length changed when lowercased, then the
found match indexes would be inaccurate. E.g. the input 'ou' could match
'Yo[uT]ube' if lower case 'Y' is 2 characters. The inaccurate indexes
could also be out-of-bounds; e.g. 'be' would match 'YouTub[e ]', which
would cause `DCHECK` crashes if `DCHECk`s are enabled.

This bug occurred for at least the HUP, HQP, and bookmark providers; but
probably all providers.

This CL updates classification to early exit when either the input or
match texts' lengths change when lowercased. This will result matches
without bolding (or non-bolding for search matches). But the match will
still have correct URL-styling or dim-styling. The lack of bold-styling
is better than incorrect bold-styling.

It's possible a page title or user input with a mix of multiple
languages could have the same length in original and lower case while
still containing problematic characters. This can occur if it contains
both characters that increase in length when lowercased, and characters
that decrease in length when lowercased. This seems rare enough to not
worry about. It doesn't seem worth it to lower case each character in
the user input and match text individually and verify they don't change
in length. Besides, the consequence of this situation would be
inaccurate bolding, not crashes.

Bug: 360873965
Change-Id: I053b0071eee91c41f4107ab60686464c76618f2d
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6642550
Reviewed-by: Ananya Seelam <ananyaseelam@google.com>
Auto-Submit: manuk hovanesian <manukh@chromium.org>
Commit-Queue: Ananya Seelam <ananyaseelam@google.com>
Cr-Commit-Position: refs/heads/main@{#1474348}
---

--- a/components/omnibox/browser/autocomplete_match_classification.cc
+++ b/components/omnibox/browser/autocomplete_match_classification.cc
@@ -4,9 +4,13 @@
 
 #include "autocomplete_match_classification.h"
 
+#include <string>
+#include <string_view>
+
 #include "base/i18n/case_conversion.h"
 #include "base/strings/string_util.h"
-#include "base/strings/utf_string_conversions.h"
+#include "components/omnibox/browser/autocomplete_match.h"
+#include "components/omnibox/browser/in_memory_url_index_types.h"
 #include "components/omnibox/browser/scored_history_match.h"
 #include "in_memory_url_index_types.h"
 
@@ -54,6 +58,20 @@
   std::u16string find_text_str = clean(find_text);
   std::u16string text_str = clean(text);
 
+  // Some international characters become multiple characters when converting
+  // case. E.g. Armenian և is 1 character lowercase, but 2 characters Եվ
+  // uppercased. Turkish has examples of the opposite, where the lowercasing can
+  // increase length. If the string length changes when lowercased, term match
+  // indexes will be off when being used to style the original-cased `text`.
+  // This will cause either `DCHECK` crashes if the incorrect index is out of
+  // bounds; or incorrect styling if the incorrect index remains in bounds or
+  // `DCHECK`s are disabled. E.g. input 'ou' would bold 'Yo[uT]ube' if lower
+  // case 'Y' is 2 characters.
+  if (find_text_str.size() != find_text.size() ||
+      text_str.size() != text.size()) {
+    return {};
+  }
+
   if (find_text_str.empty()) {
     return {};
   }
